src/dune_rules/rule_mode_decoder.mli

```diff
@@ -15,3 +15,9 @@ end
 
 val decode : Rule.Mode.t Dune_lang.Decoder.t
 val field : Rule.Mode.t Dune_lang.Decoder.fields_parser
+
+(** [is_ignored mode ~until_clean] will return if rule with [mode] should be
+    ignored. [until_clean] is used to set if [(promote (until-clean))] is
+    ignored as considerd by this function. Old versions of dune would
```

Comment:
There is a typo here, but asides from this this makes it much easier to understand how this parameter should be used. Thanks.

(This might possibly be a case where a default argument could make sense, since in new code we don't want to use `` `Ignore`` I assume?)

Reply:
I actually think we should consider just removing the old behavior and making the breaking change. The reason is that preserving the old behavior makes it impossible to build old projects without modifying the source directory. This is quite a pain for those packaging software that uses dune.

Reply:
I vaguely remember a plan to deprecate the other promote modes, would it be possible to interpret them as `(mode promote)` from now on?

Reply:
The only one that is deprecated is `patch_back_source_tree`

Reply:
@rgrinberg You mean, without patching the dune-lang version to whatever will include this change? Maybe worth asking @glondu if this would be helpful enough to break compatibility.

Reply:
> You mean, without patching the dune-lang version to whatever will include this change?

Yes. Especially given that updating the dune version in the project file might run into other errors, patching isn't a reasonable workaround.

Reply:
You might be right, as probably nobody depends on the existing incorrect behaviour.

Reply:
Unfortunately, [they do](https://github.com/ocaml/dune/pull/8512#issuecomment-1694431684) 😂 

In any case, we can just have them release a new version and then do this change.